sshd-core/src/main/java/org/apache/sshd/server/session/AbstractServerSession.java

```diff
@@ -367,9 +382,26 @@ protected String resolveAvailableSignaturesProposal(FactoryManager proposedManag
 
         KeyPairProvider kpp = getKeyPairProvider();
         boolean debugEnabled = log.isDebugEnabled();
-        Iterable<String> provided;
+        Set<String> provided = null;
         try {
-            provided = (kpp == null) ? null : kpp.getKeyTypes(this);
+            if (kpp != null) {
+                provided = StreamSupport.stream(kpp.getKeyTypes(this).spliterator(), false)
+                    .collect(Collectors.toSet());
+
+                HostKeyCertificateProvider hostKeyCertificateProvider = getHostKeyCertificateProvider();
+                if (hostKeyCertificateProvider != null) {
+                    Iterable<OpenSshCertificate> certificates = hostKeyCertificateProvider.loadCertificates(this);
+                    for (OpenSshCertificate certificate : certificates) {
+                        // Add the certificate alg only if the corresponding keyPair type is available
+                        if (provided.contains(certificate.getRawKeyType())) {
+                            provided.add(certificate.getKeyType());
+                        } else {
```

Comment:
This code interferes with a lazy-load feature (see SSHD-860) that we implemented since it consumes all available key types thus defeating the purpose of the lazy-load. We need to figure out a way to preserve that capability

Reply:
Not sure if I understand the issue. The call to `kpp.getKeyTypes(this)` did not change.
Loading all provided certificates should not be an issue and is required to get their type? (They should not be password protected since they are public.)

Reply:
The idea is that we do not want to iterate over the returned key types until really necessary - SSHD-860 implements "lazy" iterators that load keys only when 'next()' is called.  Therefore 
```java
StreamSupport.stream(kpp.getKeyTypes(this).spliterator(), false).collect(Collectors.toSet());
```
violates this "laziness" since it collects all the key types - something we wanted to avoid doing in SSHD-860

Reply:
Yes, but the call to `kpp.getKeyTypes(this)` did not change - it was there already before. I iterate only over an already existing `Set` of `String`s. That doesn't change anything regarding this lazy loading? Please correct me, if I am mistaken..

Reply:
I have reviewed the code and reached the conclusion that you are right - "laziness" is not an issue in this case